lua/nvim-treesitter/tsrange.lua

```diff
@@ -0,0 +1,145 @@
+function TSRange.from_nodes(buf, start_node, end_node)
+  TSRange.__index = TSRange
+  local start_pos = {start_node:start()}
+  local end_pos = {end_node:end_()}
+  return setmetatable(
+    {
+      start_pos = {start_pos[1], start_pos[2], start_pos[3]},
+      end_pos = {end_pos[1], end_pos[2], end_pos[3]},
+      buf = buf,
+      [1] = start_pos[1],
+      [2] = start_pos[2],
+      [3] = end_pos[1],
+      [4] = end_pos[2],
+    },
+    TSRange)
+end
+
+function TSRange.from_table(buf, range)
+  return setmetatable(
+    {
+      start_pos = {range[1], range[2], get_byte_offset(buf, range[1], range[2])},
+      end_pos = {range[3], range[4], get_byte_offset(buf, range[3], range[4])},
+      buf = buf,
+      [1] = range[1],
+      [2] = range[2],
+      [3] = range[3],
+      [4] = range[4],
+    },
+    TSRange)
+end
+
+function TSRange:parent(range)
+  local parser = parsers.get_parser(self.buf, parsers.get_buf_lang(range))
+  local root = parser.tree:root()
+  return root:named_descendant_for_range(self.start_pos[1], self.start_pos[2], self.end_pos[1], self.end_pos[2])
+end
+
+function TSRange:field()
+end
+
+function TSRange:child_count()
+  return #self:collect_children()
+end
+
+function TSRange:named_child_count()
+  return #self:collect_children(function(c) return c:named() end)
+end
+
+function TSRange:iter_children()
+  local raw_iterator = self:parent().iter_children()
+  return function()
+    while true do
+      local node = raw_iterator()
+      if not node then return end
+      local _, _, start_byte = node:start()
+      local _, _, end_byte = node:end_()
+      if start_byte >= self.start_pos[3] and end_byte <= self.end_pos[3] then
+        return node
+      end
+    end
+  end
+end
+
+function TSRange:collect_children(filter_fun)
+  local children = {}
+  for _, c in self:iter_children() do
+    if not filter_fun or filter_fun(c) then
+      table.insert(children, c)
+    end
+  end
+  return children
+end
+
+function TSRange:child(index)
+  return self:collect_children()[index + 1]
+end
+
+function TSRange:named_child(index)
+  return self:collect_children(function(c) return c.named() end)[index + 1]
+end
+
+function TSRange:start()
+  return unpack(self.start_pos)
+end
+
+function TSRange:end_()
+  return unpack(self.end_pos)
+end
+
+function TSRange:range()
+  return self.start_pos[1], self.start_pos[2], self.end_pos[1], self.end_pos[2]
+end
+
+function TSRange:type()
+  return "nvim-treesitter-virtual-node"
+end
+
+function TSRange:symbol()
+  return error('"nvim-treesitter-virtual-node" is not a real tree-sitter node!')
+end
+
+function TSRange:named()
+  return false
+end
+
+function TSRange:missing()
+  return false
+end
+
+function TSRange:has_error()
+  return #self:collect_children(function(c) return c:has_error() end) > 0 and true or false
+end
+
+function TSRange:sexpr()
+  return table.concat(vim.tbl_map(function(c) return c:sexpr() end, self:collect_children()), ' ')
+end
+
+M.TSRange = TSRange
+return M
```

Comment:
LGTM, but are methods like range:named,missing and such not implemented yet ?

Reply:
What do you mean? `TSRange` as such is neither a named nodes nor a missing node. So I just returned `false`

Reply:
I think this is not worth adding then.

Reply:
but when these methods are called and missing, it will cause an error in code that assumes mostly real nodes (usually we have only code that is aware of queries using virtual nodes)

Reply:
better to have an error than undefined behavior imo

Reply:
alright, then I remove it. Though stating that this is not a filler node or a named node is for me the opposite of defined behavior.

Reply:
well this is not a node, its a range of nodes right ?

Reply:
I think these methods shouldn't throw errors... If I wanted to use this in place of a node for I now need to ALWAYS wrap these calls in a `pcall` because there is no guarantee that it's not a virtual node. IMO it should be a noop or return false. Can this behavior be configurable? This makes it kind of hard to use in a plugin... The convenience is kind of lost.

Reply:
I would also prefer to never throw an error. those methods are for selecting special kinds of nodes. We can return with confidence that we are not one of the special nodes the user is asking for.

Reply:
well i agree, i didn't understood the purpose of this at first, but it indeed integrates smoothly with nodes.